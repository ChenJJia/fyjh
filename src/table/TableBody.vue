<template>
  <tbody>
    <tr v-for="(line, idx) in data" :key="idx">
      <td v-for="(col, index) in tableColumn" :key="index">
        {{ renderText(line, col) }}
      </td>
    </tr>
  </tbody>
</template>

<script>
export default {
    name: 'TableBody',
    props: {
        tableColumn: {
            type: Array,
            default: () => [],
        },
        data: {
            type: Array,
            default: () => [],
        },
    },
    methods: {
        renderText(line, col) {
            if (col.render) {
                return col.render(col)
            }
            return line[col.index]
        },
    },
}
</script>

<style scoped lang="less">
th, td {
    border: 1px solid #333;
}
</style>
