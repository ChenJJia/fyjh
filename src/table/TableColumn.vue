<template></template>
<script>
export default {
    name: 'TableColumn',
    props: {
        index: {
            type: String,
            default: ''
        },
        defaultWidth: {
            type: Number,
            default: 100
        },
        label: {
            type: String,
            default: ''
        }
    }
}
</script>

<style scoped lang="less">

</style>
