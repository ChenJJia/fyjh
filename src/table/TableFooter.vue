<template>
  <div v-if="curShowColNum">
    <ul class="pagine-list">
      <li
        v-for="idx in curShowColNum"
        :key="idx"
        class="pagine-item page-index"
        :class="{active: idx === curActive}"
        @click="onClickPage(idx)"
      >
        {{ idx }}
      </li>
      <template v-if="hasEllipse">
        <li class="pagine-item" @click="expandPagine">
          ...
        </li>
        <li
          class="pagine-item page-index"
          :class="{active: totalPagineNum === curActive}"
          @click="onClickPage(totalPagineNum)"
        >
          {{ totalPagineNum }}
        </li>
      </template>
    </ul>
    <span style="margin-left: 16px">共{{ tableTotal }}项</span>
  </div>
</template>

<script>
export default {
    name: 'TableFooter',
    props: {
        data: {
            type: Array,
            default: () => [],
        },
        total: {
            type: Number,
            default: 0,
        },
        local: {
            type: Boolean,
            default: false,
        },
        page: {
            type: Number,
            default: 1,
        },
        limit: {
            type: Number,
            default: 10,
        },
        showPageColNum: {
            type: Number,
            default: 3,
        },
    },
    data() {
        return {
            curActive: 1,
            curShowColNum: this.showPageColNum,
            showEllipse: true,
        }
    },
    computed: {
        tableTotal() {
            return this.local ? this.data.length : this.total
        },
        hasEllipse() {
            return this.showEllipse && this.tableTotal > (this.showPageColNum + 1) * this.limit
        },
        totalPagineNum() {
            return Math.ceil(this.tableTotal / this.limit)
        },
    },
    mounted() {
        this._initData()
    },
    methods: {
        _initData() {
            const total =  this.totalPagineNum,
                showTotal = this.showPageColNum
            this.curShowColNum = (total <= showTotal + 1) ? total : showTotal
        },
        expandPagine() {
            this.showEllipse = false
            this.curShowColNum = this.totalPagineNum
        },
        onClickPage(val) {
            this.$emit('changepage', val, this.curActive, this.limit)
            this.curActive = val
        },
        setActive (val) {
            this.curActive = val
        },
    },
}
</script>

<style scoped lang="less">
.pagine-list {
    display: inline-block;
    margin-top: 32px;
    .pagine-item {
        display: inline-block;
        cursor: pointer;
        &.page-index {
            border: 1px solid #ddd;
            height: 24px;
            width: 24px;
            text-align: center;
            &.active {
                color: #fff;
                background-color: #0080ff;
            }
        }
    }
}
</style>
