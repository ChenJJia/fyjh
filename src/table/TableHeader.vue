<template>
  <thead class="table-header">
    <tr>
      <th v-for="(col, index) in tableColumn" :key="index">
        {{ col.label }}
        <button class="table-header__btn btn-asc" @click="onSort(col, 'asc')">顺序</button>
        <button class="table-header__btn btn-des" @click="onSort(col, 'des')">倒叙</button>
        <button class="table-header__btn btn-def" @click="onSort(col)">恢复</button>
      </th>
    </tr>
  </thead>
</template>

<script>

export default {
    name: 'TableHeader',
    props: {
        tableColumn: {
            type: Array,
            default: () => [],
        },
        sortable: {
            type: Boolean,
            default: false,
        },
    },
    computed: {},
    methods: {
        onSort: function (item, order) {
            this.$emit('sort', item.index, order)
        },
    },
}
</script>
<style scoped lang="less">
.table-header {
  th, tr {
    border: 1px solid #333;
  }
  &__btn {
    cursor: pointer;
    font-size: 12px;
    border: 1px solid black;
    border-radius: 4px;
    padding: 0 2px;
    margin-right: 4px;
  }
}

</style>
