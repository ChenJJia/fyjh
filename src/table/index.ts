/**
 * Created by uedc on 2021/10/11.
 */

import TestTable from './Table.vue'
import TableColumn from './TableColumn.vue'

export { TestTable, TableColumn }
