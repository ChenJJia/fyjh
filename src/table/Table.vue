<template>
  <div>
    <table :width="width">
      <slot />
      <colgroup>
        <col v-for="col in tableColumn" :key="col.index" :name="col.index">
      </colgroup>
      <table-header :table-column="tableColumn" @sort="onSort" />
      <table-body :data="actualData" :table-column="tableColumn" />
    </table>
    <table-footer
      ref="tableFooter"
      :data="data"
      :show-page-col-num="showPageColNum"
      :limit="limit"
      :local="local"
      @changepage="changePage"
    />
  </div>
</template>
<script>

/**
 * 表格组件
 * 1、支持表头功能、表格主要内容、分页功能（对分页栏没有任何要求，可以点击并联动即可；
 * 2、支持静态数据的加载（支持本地的静态分页）；
 * 3、支持排序（顺序、倒叙、以及恢复原来的状态）；
 * 4、对css样式没有强的要求，可看即可；
 * demo：https://blog.csdn.net/weixin_44897255/article/details/105808987；
 * https://baijiahao.baidu.com/s?id=1711219715683083621&wfr=spider&for=pc
 *
 * 思路：
 * 把util放util里面（参考element，可以把util搞过来）；
 * 表格使用方式有两种：
 * 1、table+table-column：table的默认slot中放table-column，这样table内部就可以拿到所有的table-column，并将这个去转换成对应的配置项options；
 * 2、table+tableOptions：表格通过配置项去渲染表头、表格内容、分页；
 * 组件具体设计：
 * 1、使用table搭架子，然后内部分别有三个组件table-header、table-body、table-footer；
 * 2、table-header：通过获取的options的配置项生成表格头；
 * 3、table-body：表格的列通过options循环，行通过数据循环（如果数据传递太复杂，可以使用vuex）；
 * 4、table-footer：放一个分页组件pagination；
 * 5、pagination：可设置最多显示多少个，多余的显示...，还有跳转第1页和最后一页的按钮；
 * 6、动态加载数据：在options里面配置load回掉，表格会自动加载load，获取里面的数据再调用静态加载数据的方法渲染界面；
 * 7、本地的静态分页：要点就是不发送请求，界面的更新和数据获取都是自己去处理data，截取到对应的数据就行了，然后调用静态加载数据渲染；
 * 8、顺序、倒叙、恢复：options或table-column里面可配置是否可以排序sortable，如果可以排序可以给当前列配置一个sortFn去处理data，
 *   如果没配置sortFn就用默认的，根据文字排序；因为需要恢复，所以需要对数据进行备份；
 * 9、样式随便写吧；
 */
import TableHeader from './TableHeader.vue'
import TableBody from './TableBody.vue'
import TableFooter from './TableFooter.vue'
export default {
    name: 'Table',
    components: {
        TableFooter,
        TableBody,
        TableHeader,
    },
    props: {
        options: {
            type: Object,
            default: () => ({}),
        },
        width: {
            type: Number,
            default: 500,
        },
    },
    data () {
        return {
            slotColumn: [],
            actualData: this.options.data,
            currentData: this.options.data,
            oldData: this.options.data,
        }
    },
    computed: {
        tableColumn() {
            return this.options.tableColumn || this.slotColumn || []
        },
        data() {
            return this.options.data || []
        },
        showPageColNum() {
            return this.options.showPageColNum
        },
        limit() {
            return this.options.limit
        },
        local() {
            return this.options.local
        },

    },
    mounted() {
        this._initData()
    },
    methods: {
        _initData() {
            this.oldData = this.data
            this.currentData = this.data

            this.getSlotColumn()
            this._setPageNumData(1, this.limit)
        },
        getSlotColumn () {
            const { default: column = [] } = this.$slots
            this.slotColumn = column.map(col => col.componentOptions.propsData)
        },

        onSort (index, order) {
            const item = this.tableColumn.filter(item => item.index === index)?.[0]
            if (!order) {
                this.currentData = [...this.oldData]
            } else {
                this.currentData = [...this.oldData].sort((item1, item2) => {
                    if (item.sortFn) {
                        return item.sortFn(item1, item2)
                    }
                    return item1[index] > item2[index] ? -1 : 1
                })
                order === 'asc' && this.currentData.reverse()
            }
            this.$refs.tableFooter.setActive(1)
            this._setPageNumData(1, this.limit)
        },
        changePage(val, pre, limit) {
            this.$emit('changepage', val, pre, limit)
            this._setPageNumData(val, limit)
        },
        _setPageNumData(page, limit) {
            this.actualData = this.currentData.slice((page-1)*limit, page*limit)
        },
    },
}
</script>
<style>
table {
    border: 1px solid #333;
    border-collapse: collapse;
}
</style>
