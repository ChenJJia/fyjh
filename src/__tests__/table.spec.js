import { mount } from '@vue/test-utils'
import { TestTable } from '../table'

describe('Table', () => {
    const TableMount = options => mount(TestTable, options)

    test('render', () => {
        const wrapper = TableMount()
        // expect(wrapper.html()).toMatchSnapshot()
        expect(() => {
            wrapper.vm.$forceUpdate()
            wrapper.vm.$destroy()
        }).not.toThrow()
    })

    test('click sort', async () => {
        const options = {
            propsData: {
                options: {
                    data: [
                        { name: 'cjj', age: 18 },
                        { name: 'we', age: 32 },
                        { name: 'r', age: 43 },
                        { name: 'ew', age: 34 },
                        { name: 'fds', age: 5 },
                        { name: 'd', age: 43 },
                        { name: 'f', age: 5 },
                        { name: 'fzf', age: 23 },
                    ],
                    tableColumn: [
                        { label: '名字', index: 'name', sortFn: (item1, item2) => {
                            return item1.name > item2.name ? -1 : 1
                        }  },
                        { label: '年龄', index: 'age' },
                    ],
                    local: true,
                    limit: 2,
                    showPageColNum: 2,
                },
            },
        }
        const wrapper = TableMount(options)
        await wrapper.find('button.btn-asc').trigger('click')
        expect(wrapper.findComponent({ name: 'TableBody' }).vm.data.map(item => item.name).join('')).toBe('cjjd')

        await wrapper.find('button.btn-des').trigger('click')
        expect(wrapper.findComponent({ name: 'TableBody' }).vm.data.map(item => item.name).join('')).toBe('wer')

        await wrapper.find('button.btn-def').trigger('click')
        expect(wrapper.findComponent({ name: 'TableBody' }).vm.data.map(item => item.name).join('')).toBe('cjjwe')

        await wrapper.find('li.active').trigger('click')
    })

    // test('props test', () => {
    //     const wrapper = TableMount({
    //         propsData: {
    //             test: true,
    //         },
    //     })
    //
    //     expect(wrapper.find('.test-class').exists()).toBeTruthy()
    // })
})
