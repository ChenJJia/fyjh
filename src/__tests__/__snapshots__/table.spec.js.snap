// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<table width=\\"500\\">
  <colgroup></colgroup>
  <thead>
    <tr></tr>
  </thead>
  <tbody></tbody>
  <div>
    <ul class=\\"pagine-list\\">
      <li class=\\"pagine-item page-index active\\">
        1
      </li>
      <li class=\\"pagine-item page-index\\">
        2
      </li>
      <li class=\\"pagine-item page-index\\">
        3
      </li>
      <!---->
    </ul> <span style=\\"margin-left: 16px;\\">共0项</span>
  </div>
</table>"
`;
